File: swift/lib/Sources/Proxy/PlatformProxy.swift
/// This carries no mutable state of its own and reaches only registries that guard theirs.
class PlatformProxy: ObjCPlatformProxyImpl, @unchecked Sendable {
    static let shared: PlatformProxy = .init()
    override private init() {}

    override func onNativeProxyCallback(_ key: String, _ data: String) {
        CallbackList.shared.execute(key: key, data: data)
    }

    override func onNativeProxyCall(_ key: String, _ data: String) {
        guard let config = XPLPC.shared.config else {
            Log.e("[PlatformProxy : call] XPLPC was not initialized")
            callNativeProxyCallback(key, "")
            return
        }

        guard let request = config.serializer.decodeRequest(data) else {
            callNativeProxyCallback(key, "")
            return
        }

        if request.functionName.isEmpty {
            Log.e("[PlatformProxy : call] Function name is empty")
            callNativeProxyCallback(key, "")
            return
        }

        guard let mappingItem = MappingList.shared.find(request.functionName) else {
            Log.e("[PlatformProxy : call] Mapping not found for function: \(request.functionName)")
            callNativeProxyCallback(key, "")
            return
        }

        mappingItem.target(request.message) { [weak self] response in
            guard let self else {
                Log.e("[PlatformProxy : call] The proxy is gone, so this answer is lost")
                return
            }

            self.callNativeProxyCallback(key, config.serializer.encodeFunctionReturnValue(response))
        }
    }

    override func onHasMapping(_ name: String) -> Bool {
        return MappingList.shared.has(name)
    }

    override func onInitializePlatform() {}

    override func onFinalizePlatform() {
        MappingList.shared.clear()
    }
}

File: kotlin/lib/library/src/main/java/com/xplpc/proxy/PlatformProxy.kt
package com.xplpc.proxy

import com.xplpc.core.XPLPC
import com.xplpc.data.CallbackList
import com.xplpc.data.MappingList
import com.xplpc.message.Message
import com.xplpc.util.Log

class PlatformProxy {
    companion object {
        @JvmStatic
        external fun nativeProxyCall(key: String, data: String)

        @JvmStatic
        external fun nativeCallProxyCallback(key: String, data: String)

        @JvmStatic
        fun onCallProxyCallback(key: String, data: String) {
            CallbackList.execute(key, data)
        }

        @JvmStatic
        fun onNativeProxyCall(key: String, data: String) {
            // function name
            val functionName = XPLPC.config.serializer.decodeFunctionName(data)

            if (functionName.isEmpty()) {
                Log.e("[PlatformProxy : call] Function name is empty")
                nativeCallProxyCallback(key, data)
                return
            }

            // mapping item
            val mappingItem = MappingList.find(functionName)

            if (mappingItem == null) {
                Log.e("[PlatformProxy : call] Mapping not found for function: $functionName")
                nativeCallProxyCallback(key, data)
                return
            }

            // message
            var message: Message? = null

            try {
                message = XPLPC.config.serializer.decodeMessage(data)
            } catch (e: Exception) {
                Log.e("[PlatformProxy : call] Error when decode message: ${e.message}")
            }

            if (message == null) {
                Log.e("[PlatformProxy : call] Error when decode message for function: $functionName")
                nativeCallProxyCallback(key, data)
                return
            }

            // execute
            mappingItem.target(message) { response ->
                val encodedData: String?

                try {
                    encodedData = XPLPC.config.serializer.encodeFunctionReturnValue(response)
                } catch (e: Exception) {
                    Log.e("[PlatformProxy : call] Error when encode data: ${e.message}")
                    nativeCallProxyCallback(key, data)
                    return@target
                }

                nativeCallProxyCallback(key, encodedData)
            }
        }
    }
}
